set log level via pd-ctl succeed, but don't work.

## Bug Report

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What did you do?
/var/lib/pd/log # /pd-ctl log debug
Success!
/var/lib/pd/log # tail -f pd.log
[2021/12/28 05:59:39.271 +00:00] [ERROR] [client.go:171] ["region sync with leader meet error"] [error="[PD:grpc:ErrGRPCRecv]rpc error: code = Unavailable desc = server not started"]
[2021/12/28 05:59:40.272 +00:00] [INFO] [client.go:167] ["server starts to synchronize with leader"] [server=tc-pd-0] [leader=tc-pd-2] [request-index=0]
[2021/12/28 06:32:40.747 +00:00] [WARN] [grpclog.go:60] ["grpc: Server.processUnaryRPC failed to write status: connection error: desc = \"transport is closing\""]
[2021/12/28 06:34:41.288 +00:00] [WARN] [grpclog.go:60] ["grpc: Server.processUnaryRPC failed to write status: connection error: desc = \"transport is closing\""]
[2021/12/28 06:43:58.963 +00:00] [WARN] [grpclog.go:60] ["transport: http2Server.HandleStreams failed to read frame: read tcp 10.233.86.141:2379->10.233.116.155:59226: read: connection reset by peer"]
[2021/12/28 06:48:27.824 +00:00] [WARN] [grpclog.go:60] ["grpc: Server.processUnaryRPC failed to write status: connection error: desc = \"transport is closing\""]
[2021/12/28 06:59:00.694 +00:00] [INFO] [index.go:189] ["compact tree index"] [revision=1]
[2021/12/28 06:59:00.695 +00:00] [INFO] [kvstore_compaction.go:55] ["finished scheduled compaction"] [compact-revision=1] [took=34.19µs]
[2021/12/28 07:59:00.707 +00:00] [INFO] [index.go:189] ["compact tree index"] [revision=2414]
[2021/12/28 07:59:00.734 +00:00] [INFO] [kvstore_compaction.go:55] ["finished scheduled compaction"] [compact-revision=2414] [took=26.781576ms]

<!-- If possible, provide a recipe for reproducing the error. -->

### What did you expect to see?

### What did you see instead?

### What version of PD are you using (`pd-server -V`)?
/ # /pd-server -V
Release Version: v5.4.0-nightly
Edition: Community
Git Commit Hash: ea57c6d6a0c9ad081aded76545f3d97bc465635f
Git Branch: heads/refs/tags/v5.4.0-nightly
UTC Build Time:  2021-12-28 04:09:10

--- a/tests/pdctl/log/log_test.go
+++ b/tests/pdctl/log/log_test.go
@@ -94,3 +94,60 @@ func (s *logTestSuite) TestLog(c *C) {
 		c.Assert(svr.GetConfig().Log.Level, Equals, testCase.expect)
 	}
 }
+
+func (s *logTestSuite) TestInstanceLog(c *C) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	cluster, err := tests.NewTestCluster(ctx, 3)
+	c.Assert(err, IsNil)
+	err = cluster.RunInitialServers()
+	c.Assert(err, IsNil)
+	cluster.WaitLeader()
+	pdAddrs := cluster.GetConfig().GetClientURLs()
+	cmd := pdctlCmd.GetRootCmd()
+
+	store := &metapb.Store{
+		Id:            1,
+		State:         metapb.StoreState_Up,
+		LastHeartbeat: time.Now().UnixNano(),
+	}
+	leaderServer := cluster.GetServer(cluster.GetLeader())
+	c.Assert(leaderServer.BootstrapCluster(), IsNil)
+	svr := leaderServer.GetServer()
+	pdctl.MustPutStore(c, svr, store)
+	defer cluster.Destroy()
+
+	var testCases = []struct {
+		cmd      []string
+		instance string
+		expect   string
+	}{
+		// log [fatal|error|warn|info|debug] [address]
+		{
+			cmd:      []string{"-u", pdAddrs[0], "log", "debug", pdAddrs[0]},
+			instance: pdAddrs[0],
+			expect:   "debug",
+		},
+		{
+			cmd:      []string{"-u", pdAddrs[0], "log", "error", pdAddrs[1]},
+			instance: pdAddrs[1],
+			expect:   "error",
+		},
+		{
+			cmd:      []string{"-u", pdAddrs[0], "log", "warn", pdAddrs[2]},
+			instance: pdAddrs[2],
+			expect:   "warn",
+		},
+	}
+
+	for _, testCase := range testCases {
+		_, err = pdctl.ExecuteCommand(cmd, testCase.cmd...)
+		c.Assert(err, IsNil)
+		svrs := cluster.GetServers()
+		for _, svr := range svrs {
+			if svr.GetAddr() == testCase.instance {
+				c.Assert(svr.GetConfig().Log.Level, Equals, testCase.expect)
+			}
+		}
+	}
+}

--- a/tests/pdctl/log/log_test.go
+++ b/tests/pdctl/log/log_test.go
@@ -33,90 +33,79 @@ func Test(t *testing.T) {
 
 var _ = Suite(&logTestSuite{})
 
-type logTestSuite struct{}
+type logTestSuite struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	cluster *tests.TestCluster
+	svr     *server.Server
+	pdAddrs []string
+}
 
 func (s *logTestSuite) SetUpSuite(c *C) {
 	server.EnableZap = true
-}
-
-func (s *logTestSuite) TestLog(c *C) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	cluster, err := tests.NewTestCluster(ctx, 1)
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	var err error
+	s.cluster, err = tests.NewTestCluster(s.ctx, 3)
 	c.Assert(err, IsNil)
-	err = cluster.RunInitialServers()
+	err = s.cluster.RunInitialServers()
 	c.Assert(err, IsNil)
-	cluster.WaitLeader()
-	pdAddr := cluster.GetConfig().GetClientURL()
-	cmd := pdctlCmd.GetRootCmd()
+	s.cluster.WaitLeader()
+	s.pdAddrs = s.cluster.GetConfig().GetClientURLs()
 
 	store := &metapb.Store{
 		Id:            1,
 		State:         metapb.StoreState_Up,
 		LastHeartbeat: time.Now().UnixNano(),
 	}
-	leaderServer := cluster.GetServer(cluster.GetLeader())
+	leaderServer := s.cluster.GetServer(s.cluster.GetLeader())
 	c.Assert(leaderServer.BootstrapCluster(), IsNil)
-	svr := leaderServer.GetServer()
-	pdctl.MustPutStore(c, svr, store)
-	defer cluster.Destroy()
+	s.svr = leaderServer.GetServer()
+	pdctl.MustPutStore(c, s.svr, store)
+}
 
+func (s *logTestSuite) TearDownSuite(c *C) {
+	s.cluster.Destroy()
+	s.cancel()
+}
+
+func (s *logTestSuite) TestLog(c *C) {
+	cmd := pdctlCmd.GetRootCmd()
 	var testCases = []struct {
 		cmd    []string
 		expect string
 	}{
 		// log [fatal|error|warn|info|debug]
 		{
-			cmd:    []string{"-u", pdAddr, "log", "fatal"},
+			cmd:    []string{"-u", s.pdAddrs[0], "log", "fatal"},
 			expect: "fatal",
 		},
 		{
-			cmd:    []string{"-u", pdAddr, "log", "error"},
+			cmd:    []string{"-u", s.pdAddrs[0], "log", "error"},
 			expect: "error",
 		},
 		{
-			cmd:    []string{"-u", pdAddr, "log", "warn"},
+			cmd:    []string{"-u", s.pdAddrs[0], "log", "warn"},
 			expect: "warn",
 		},
 		{
-			cmd:    []string{"-u", pdAddr, "log", "info"},
+			cmd:    []string{"-u", s.pdAddrs[0], "log", "info"},
 			expect: "info",
 		},
 		{
-			cmd:    []string{"-u", pdAddr, "log", "debug"},
+			cmd:    []string{"-u", s.pdAddrs[0], "log", "debug"},
 			expect: "debug",
 		},
 	}
 
 	for _, testCase := range testCases {
-		_, err = pdctl.ExecuteCommand(cmd, testCase.cmd...)
+		_, err := pdctl.ExecuteCommand(cmd, testCase.cmd...)
 		c.Assert(err, IsNil)
-		c.Assert(svr.GetConfig().Log.Level, Equals, testCase.expect)
+		c.Assert(s.svr.GetConfig().Log.Level, Equals, testCase.expect)
 	}
 }
 
 func (s *logTestSuite) TestInstanceLog(c *C) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	cluster, err := tests.NewTestCluster(ctx, 3)
-	c.Assert(err, IsNil)
-	err = cluster.RunInitialServers()
-	c.Assert(err, IsNil)
-	cluster.WaitLeader()
-	pdAddrs := cluster.GetConfig().GetClientURLs()
 	cmd := pdctlCmd.GetRootCmd()
-
-	store := &metapb.Store{
-		Id:            1,
-		State:         metapb.StoreState_Up,
-		LastHeartbeat: time.Now().UnixNano(),
-	}
-	leaderServer := cluster.GetServer(cluster.GetLeader())
-	c.Assert(leaderServer.BootstrapCluster(), IsNil)
-	svr := leaderServer.GetServer()
-	pdctl.MustPutStore(c, svr, store)
-	defer cluster.Destroy()
-
 	var testCases = []struct {
 		cmd      []string
 		instance string
@@ -124,26 +113,26 @@ func (s *logTestSuite) TestInstanceLog(c *C) {
 	}{
 		// log [fatal|error|warn|info|debug] [address]
 		{
-			cmd:      []string{"-u", pdAddrs[0], "log", "debug", pdAddrs[0]},
-			instance: pdAddrs[0],
+			cmd:      []string{"-u", s.pdAddrs[0], "log", "debug", s.pdAddrs[0]},
+			instance: s.pdAddrs[0],
 			expect:   "debug",
 		},
 		{
-			cmd:      []string{"-u", pdAddrs[0], "log", "error", pdAddrs[1]},
-			instance: pdAddrs[1],
+			cmd:      []string{"-u", s.pdAddrs[0], "log", "error", s.pdAddrs[1]},
+			instance: s.pdAddrs[1],
 			expect:   "error",
 		},
 		{
-			cmd:      []string{"-u", pdAddrs[0], "log", "warn", pdAddrs[2]},
-			instance: pdAddrs[2],
+			cmd:      []string{"-u", s.pdAddrs[0], "log", "warn", s.pdAddrs[2]},
+			instance: s.pdAddrs[2],
 			expect:   "warn",
 		},
 	}
 
 	for _, testCase := range testCases {
-		_, err = pdctl.ExecuteCommand(cmd, testCase.cmd...)
+		_, err := pdctl.ExecuteCommand(cmd, testCase.cmd...)
 		c.Assert(err, IsNil)
-		svrs := cluster.GetServers()
+		svrs := s.cluster.GetServers()
 		for _, svr := range svrs {
 			if svr.GetAddr() == testCase.instance {
 				c.Assert(svr.GetConfig().Log.Level, Equals, testCase.expect)

--- a/tools/pd-ctl/pdctl/command/global.go
+++ b/tools/pd-ctl/pdctl/command/global.go
@@ -162,13 +162,7 @@ func getEndpoints(cmd *cobra.Command) []string {
 		cmd.Println("get pd address failed, should set flag with '-u'")
 		os.Exit(1)
 	}
-	eps := strings.Split(addrs, ",")
-	for i, ep := range eps {
-		if j := strings.Index(ep, "//"); j == -1 {
-			eps[i] = "//" + ep
-		}
-	}
-	return eps
+	return strings.Split(addrs, ",")
 }
 
 func postJSON(cmd *cobra.Command, prefix string, input map[string]interface{}) {
@@ -232,6 +226,9 @@ func doGet(endpoint, prefix, method string, resp *string, customHeader http.Head
 }
 
 func checkURL(endpoint string) (string, error) {
+	if j := strings.Index(endpoint, "//"); j == -1 {
+		endpoint = "//" + endpoint
+	}
 	var u *url.URL
 	u, err := url.Parse(endpoint)
 	if err != nil {

--- a/tools/pd-ctl/pdctl/command/log_command.go
+++ b/tools/pd-ctl/pdctl/command/log_command.go
@@ -29,7 +29,7 @@ var (
 // NewLogCommand New a log subcommand of the rootCmd
 func NewLogCommand() *cobra.Command {
 	conf := &cobra.Command{
-		Use:   "log [fatal|error|warn|info|debug]",
+		Use:   "log [fatal|error|warn|info|debug] [addr]",
 		Short: "set log level",
 		Run:   logCommandFunc,
 	}
@@ -38,7 +38,7 @@ func NewLogCommand() *cobra.Command {
 
 func logCommandFunc(cmd *cobra.Command, args []string) {
 	var err error
-	if len(args) != 1 {
+	if len(args) == 0 || len(args) > 2 {
 		cmd.Println(cmd.UsageString())
 		return
 	}
@@ -48,11 +48,26 @@ func logCommandFunc(cmd *cobra.Command, args []string) {
 		cmd.Printf("Failed to set log level: %s\n", err)
 		return
 	}
-	_, err = doRequest(cmd, logPrefix, http.MethodPost, http.Header{"Content-Type": {"application/json"}},
-		WithBody(bytes.NewBuffer(data)))
-	if err != nil {
-		cmd.Printf("Failed to set log level: %s\n", err)
-		return
+
+	if len(args) == 2 {
+		url, err := checkURL(args[1])
+		if err != nil {
+			cmd.Printf("Failed to parse address %v: %s\n", args[1], err)
+			return
+		}
+		_, err = doRequestSingleEndpoint(cmd, url, logPrefix, http.MethodPost, http.Header{"Content-Type": {"application/json"}, "PD-Allow-follower-handle": {"true"}},
+			WithBody(bytes.NewBuffer(data)))
+		if err != nil {
+			cmd.Printf("Failed to set %v log level: %s\n", args[1], err)
+			return
+		}
+	} else {
+		_, err = doRequest(cmd, logPrefix, http.MethodPost, http.Header{"Content-Type": {"application/json"}},
+			WithBody(bytes.NewBuffer(data)))
+		if err != nil {
+			cmd.Printf("Failed to set log level: %s\n", err)
+			return
+		}
 	}
 	cmd.Println("Success!")
 }

